curvefs/src/mds/schedule/leaderScheduler.cpp

```diff
@@ -0,0 +1,387 @@
+ *  limitations under the License.
+ */
+
+/*
+ * @Project: curve
+ * @Date: 2022-04-07 14:15:49
+ * @Author: chenwei
+ */
+
+#include "curvefs/src/mds/schedule/operatorFactory.h"
+#include "curvefs/src/mds/schedule/scheduler.h"
+
+namespace curvefs {
+namespace mds {
+namespace schedule {
+/**
+ * use curl -L mdsIp:port/flags/enableRapidLeaderScheduler?setvalue=true
+ * for dynamic parameter configuration
+ *
+ * If this value is set true, ignore metaserverCoolingTimeSec_
+ * when choose metaserver as a leader when leader schedule.
+ *
+ * if this value is set true, when no leader schedule operation at this round,
+ * set the value to false automaticlly.
+ */
+static bool pass_bool(const char *, bool) { return true; }
+DEFINE_bool(enableRapidLeaderScheduler, false, "rapid leader scheduler once");
+DEFINE_validator(enableRapidLeaderScheduler, &pass_bool);
+
+int LeaderScheduler::Schedule() {
+    LOG(INFO) << "LeaderScheduler begin";
+
+    int oneRoundGenOp = 0;
+    for (auto poolId : topo_->Getpools()) {
+        oneRoundGenOp += LeaderSchedulerForPool(poolId);
+    }
+
+    LOG(INFO) << "LeaderScheduler generate " << oneRoundGenOp
+              << " operators at this round";
+
+    if (FLAGS_enableRapidLeaderScheduler && oneRoundGenOp == 0) {
+        LOG(INFO) << "LeaderScheduler enable rapid leader scheduler and "
+                  << " generate no opration in this round, "
+                  << "set enableRapidLeaderScheduler to false";
+        FLAGS_enableRapidLeaderScheduler = false;
+    }
+    return oneRoundGenOp;
+}
+
+int64_t LeaderScheduler::GetRunningInterval() { return runInterval_; }
+
+int LeaderScheduler::LeaderSchedulerForPool(PoolIdType poolId) {
+    int oneRoundGenOp = 0;
+    uint16_t replicaNum = topo_->GetStandardReplicaNumInPool(poolId);
+    if (replicaNum == 0) {
+        LOG(ERROR) << "leader scheduler for pool = " << poolId
+                   << " fail, replicaNum is 0";
+        return 0;
+    }
+
+    // find metaserver with maximum and minimum number of leaders
+    double maxMoreNum = -1;
+    int maxId = -1;
+    double minLessNum = 0;
+    int minId = -1;
+    MetaServerInfo moreLeaderMetaserver;
+    MetaServerInfo lessLeaderMetaserver;
+    std::vector<MetaServerInfo> msInfos = topo_->GetMetaServersInPool(poolId);
+    static std::random_device rd;
+    static std::mt19937 g(rd());
+    std::shuffle(msInfos.begin(), msInfos.end(), g);
+
+    for (auto msInfo : msInfos) {
+        // skip unhealthy metaserver
+        if (!msInfo.IsHealthy()) {
+            continue;
+        }
+
+        double standardLeaderNum = msInfo.copysetNum * 1.0 / replicaNum;
+        // if different > 0, it means the leader num on metaserver is more than
+        // stardard leader num, vice versa
+        double different = msInfo.leaderNum - standardLeaderNum;
+
+        // if leader num is appropriate, skip this metaserver.
+        if (different > -1 && different < 1) {
```

Comment:
Does the leader balance is not absolute equilibrium and at [-1,1] is considered balance?

Reply:
yes